File: C#/A_Star/A_Star/A_Star/Grid.cs
using System.Reflection.PortableExecutable;
using System.Text;

namespace Main.A_Star;

public class Grid
{
    public Grid(String path)
    {
        String line;
        this.grid = new List<List<Spot>>();
        try
        {
            StreamReader sr = new StreamReader(path);
            line = sr.ReadLine();
            while (line != null)
            {
                this.grid.Add(new List<Spot>());
                foreach (char a in line)
                {
                    if (a != ' ')
                    {
                        this.grid.Last().Add(new Spot((int)Char.GetNumericValue(a), this.grid.Count - 1,this.grid.Last().Count));
                    }
                }
                line = sr.ReadLine();
            }
            sr.Close();
        }
        catch(Exception e)
        {
            Console.WriteLine("Exception: " + e.Message);
        }
    }

 public Grid A_star(){
        List<Spot> openSet = new List<Spot>();
        List<Spot> closedSet = new List<Spot>();
        List<Spot> path = new List<Spot>();
        Spot start = this.grid.Last().First();
        Spot end = this.grid.First().Last();
        openSet.Add(start);
        while(openSet.Any()) {
            Spot current = openSet[0];
            foreach (var spot in openSet) {
                if (spot.f < current.f) {
                    current = spot;
                }
            }
            if (current.PositionEquals(end)) {
                Spot temp = current;
                while (temp.previous != null) {
                    path.Add(temp);
                    temp = temp.previous;
                }
                for(int i = 0;i<path.Count;i++){
                    this.grid[path[i].x][path[i].y].value = 3;
                }
                this.grid[start.x][start.y].value = 3;
                this.grid[end.x][end.y].value = 3;
                return this;
            }

            openSet.Remove(current);
            if(current.y < this.grid[0].Count-1){
                current.neighbors.Add(this.grid[current.x][current.y+1]);
            }
            if (current.y  > 0) {
                current.neighbors.Add(this.grid[current.x][current.y-1]);
            }
            if (current.x > 0) {
                current.neighbors.Add(this.grid[current.x-1][current.y]);
            }
            if (current.x < this.grid.Count - 1) {
                current.neighbors.Add(this.grid[current.x+1][current.y]);
            }
            closedSet.Add(current);
            for(int i = 0;i<current.neighbors.Count;i++) {
                Spot neighbor = current.neighbors[i];
                if (neighbor.value != 5 & current.value != 5 & !includes(closedSet,neighbor)) {
                    double tempG = current.g + 1;
                    bool newPath = false;
                    if (includes(openSet,neighbor)) {
                        if (tempG < neighbor.g) {
                            neighbor.g = tempG;
                            newPath = true;
                        }
                    } else {
                        neighbor.g = tempG;
                        newPath = true;
                        openSet.Add(neighbor);
                    }
                    if (newPath) {
                        neighbor.h = heuristic(neighbor, end);
                        neighbor.f = neighbor.g + neighbor.h;
                        neighbor.previous = current;
                    }
                }
            }
        }
        Console.WriteLine("Brak ścieżki");
        return this;
        
 }
 
    double heuristic(Spot a, Spot b) {
        return Math.Sqrt(Math.Pow(a.x - b.x, 2) + Math.Pow(a.y - b.y, 2));
    }
    public bool includes(List<Spot> list, Spot spot)
    {
        for(int i = 0;i<list.Count;i++)
        {
            if (list[i] == spot)
            {
                return true;
            }
        }
        return false;
    }
    public override string ToString()
    {
        StringBuilder result = new StringBuilder();
        foreach (List<Spot> row in this.grid)
        {
            foreach (Spot spot in row)
            {
                result.Append(spot.value + " ");
            }
            result.Append( "\n");
        }
        return result.ToString();
    }

    private List<List<Spot>> _grid;
    public List<List<Spot>> grid
    {
        get => _grid;
        set => _grid = value;
    }
}

File: Java/A_Star/src/A_star/Grid.java
package A_star;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Grid {
    public Grid(String fileName) {
        this.grid = new ArrayList<>();

        List<String> lines = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = br.readLine()) != null) {
                this.grid.add(new ArrayList<>());
                for (char c : line.toCharArray()) {
                    if (c != ' ') {
                        this.grid.get(this.grid.size() - 1).add(new Spot(Character.getNumericValue(c), this.grid.size() - 1, this.grid.get(this.grid.size() - 1).size()));
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    Grid A_star(){
        ArrayList<Spot> openSet = new ArrayList<>();
        ArrayList<Spot> closedSet = new ArrayList<>();
        ArrayList<Spot> path = new ArrayList<>();
        Spot start = this.grid.get(0).get(0);
        Spot end = this.grid.get(this.grid.size() - 1).get(this.grid.get(this.grid.size() - 1).size() - 1);
        openSet.add(start);
        while (!openSet.isEmpty()) {
            Spot current = openSet.get(0);
            for (Spot spot : openSet) {
                if (spot.getF() < current.getF()) {
                    current = spot;
                }
            }
            if (current == end) {
                return this;
            }
            openSet.remove(current);
            closedSet.add(current);
            for()
            for (Spot neighbor : current.getNeighbors()) {
                if (closedSet.contains(neighbor)) {
                    continue;
                }
                double tempG = current.getG() + neighbor.getValue();
                if (!openSet.contains(neighbor)) {
                    openSet.add(neighbor);
                } else if (tempG >= neighbor.getG()) {
                    continue;
                }
                neighbor.setPrevious(current);
                neighbor.setG(tempG);
                neighbor.setH(neighbor.getManhattanDistance(end));
                neighbor.setF(neighbor.getG() + neighbor.getH());
            }
        }
        return null;
    }
    private boolean include(Spot spot, ArrayList<Spot> list){
        for (Spot s : list){
            if (s == spot){
                return true;
            }
        }
        return false;
    }

    private double heuristic(Spot spot, Spot end){
        return Math.sqrt(Math.pow(spot.getX() - end.getX(), 2) + Math.pow(spot.getY() - end.getY(), 2));
    }
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (ArrayList<Spot> row : grid) {
            for (Spot spot : row) {
                sb.append(spot.getValue() + "  ");
            }
            sb.append("\n");
        }
        return sb.toString();
    }

    private final ArrayList<ArrayList<Spot>> grid;
}
